File: app-android/src/main/java/com/ustadmobile/port/android/view/SelectSaleProductRecyclerAdapter.java
package com.ustadmobile.port.android.view;

import android.app.Activity;
import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.content.res.Resources;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.SelectSaleProductPresenter;
import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.lib.db.entities.SaleNameWithImage;

import java.io.File;

import static com.ustadmobile.lib.db.entities.SaleProductGroup.PRODUCT_GROUP_TYPE_PRODUCT;

public class SelectSaleProductRecyclerAdapter extends PagedListAdapter<SaleNameWithImage,
                SelectSaleProductRecyclerAdapter.SelectSaleProductViewHolder> {

    private static final int IMAGE_WITH = 26;
    private Context theContext;
    private Activity theActivity;
    private Fragment theFragment;
    SelectSaleProductPresenter mPresenter;

    private boolean listCategory;
    private boolean isCatalog;

    @NonNull
    @Override
    public SelectSaleProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {


        View list = LayoutInflater.from(theContext).inflate(
                R.layout.item_sale_product_blob, parent, false);
        return new SelectSaleProductViewHolder(list);

    }

    private static int dpToPxImagePerson() {
        return (int) (IMAGE_WITH
                * Resources.getSystem().getDisplayMetrics().density);
    }

    private void setPictureOnView(String imagePath, ImageView theImage) {

        Uri imageUri = Uri.fromFile(new File(imagePath));

        Picasso
                .get()
                .load(imageUri)
                .resize(dpToPxImagePerson(), dpToPxImagePerson())
                .noFade()
                .into(theImage);
    }

    @Override
    public void onBindViewHolder(@NonNull SelectSaleProductViewHolder holder, int position) {

        SaleNameWithImage entity = getItem(position);
        ImageView imageView = holder.itemView.findViewById(R.id.item_sale_product_blob_image);
        TextView name = holder.itemView.findViewById(R.id.item_sale_product_blob_title);
        ImageView dots = holder.itemView.findViewById(R.id.item_sale_product_blob_dots);

        long pictureUid = entity.getPictureUid();
        String imagePath = "";
        if (pictureUid != 0) {
            imagePath = UmAppDatabase.getInstance(theContext).getSaleProductPictureDao()
                    .getAttachmentPath(pictureUid);
        }

        if(imagePath != null && !imagePath.isEmpty())
            setPictureOnView(imagePath, imageView);
        else
            imageView.setImageResource(R.drawable.ic_card_giftcard_black_24dp);

        name.setText(entity.getName());

        if(isCatalog){
            dots.setVisibility(View.VISIBLE);
            //Options to Edit/Delete every schedule in the list
            dots.setOnClickListener((View v) -> {
                if(theActivity != null){
                    //creating a popup menu
                    PopupMenu popup = new PopupMenu(theActivity.getApplicationContext(), v);
                    popup.setOnMenuItemClickListener(item -> {
                        int i = item.getItemId();
                        if (i == R.id.edit) {
                            mPresenter.handleClickProduct(entity.getProductUid(), listCategory);
                            return true;
                        } else if (i == R.id.delete) {
                            mPresenter.handleDelteSaleProduct(entity.getProductUid(), listCategory);
                            return true;
                        } else {
                            return false;
                        }
                    });

                    //inflating menu from xml resource
                    popup.inflate(R.menu.menu_edit_delete);
                    popup.getMenu().findItem(R.id.edit).setVisible(true);
                    //displaying the popup
                    popup.show();
                }else if(theFragment != null){
                    //creating a popup menu
                    PopupMenu popup = new PopupMenu(theFragment.getContext(), v);
                    popup.setOnMenuItemClickListener(item -> {
                        int i = item.getItemId();
                        if (i == R.id.edit) {
                            mPresenter.handleClickProduct(entity.getProductUid(), listCategory);
                            return true;
                        } else if (i == R.id.delete) {
                            mPresenter.handleDelteSaleProduct(entity.getProductUid(), listCategory);
                            return true;
                        } else {
                            return false;
                        }
                    });

                    //inflating menu from xml resource
                    popup.inflate(R.menu.menu_edit_delete);
                    popup.getMenu().findItem(R.id.edit).setVisible(true);
                    //displaying the popup
                    popup.show();
                }

            });
        }else{
            dots.setVisibility(View.GONE);
        }


        holder.itemView.setOnClickListener(v -> {
            mPresenter.handleClickProduct(entity.getProductUid(), listCategory);
        });

    }

    class SelectSaleProductViewHolder extends RecyclerView.ViewHolder {
        SelectSaleProductViewHolder(View itemView) {
            super(itemView);
        }
    }

    SelectSaleProductRecyclerAdapter(
            @NonNull DiffUtil.ItemCallback<SaleNameWithImage> diffCallback,
            SelectSaleProductPresenter thePresenter,
            Activity activity,
            Boolean isCategory,
            Boolean catalog,
            Context context) {
        super(diffCallback);
        mPresenter = thePresenter;
        theActivity = activity;
        listCategory = isCategory;
        isCatalog = catalog;
        theContext = context;
    }

    SelectSaleProductRecyclerAdapter(
            @NonNull DiffUtil.ItemCallback<SaleNameWithImage> diffCallback,
            SelectSaleProductPresenter thePresenter,
            Fragment fragment,
            Boolean isCategory,
            Boolean catalog,
            Context context) {
        super(diffCallback);
        mPresenter = thePresenter;
        theFragment = fragment;
        listCategory = isCategory;
        isCategory = catalog;
        theContext = context;
    }


}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/SelectSaleProductRecyclerAdapter.kt
package com.ustadmobile.port.android.view

import android.app.Activity
import android.content.Context
import android.content.res.Resources
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.PopupMenu
import android.widget.TextView

import androidx.fragment.app.Fragment
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView

import com.squareup.picasso.Picasso
import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.SelectSaleProductPresenter
import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.db.dao.SaleProductPictureDao
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.lib.db.entities.SaleProduct
import kotlinx.coroutines.*

import java.io.File

class SelectSaleProductRecyclerAdapter
    : PagedListAdapter<SaleProduct, SelectSaleProductRecyclerAdapter.SelectSaleProductViewHolder> {
    private var theContext: Context
    private var theActivity: Activity? = null
    private var theFragment: Fragment? = null
    internal var mPresenter: SelectSaleProductPresenter

    private var listCategory: Boolean = false
    private var isCatalog: Boolean = false

    private var impl = UstadMobileSystemImpl.instance

    private var productPictureDaoRepo : SaleProductPictureDao?= null
    private var productPictureDao : SaleProductPictureDao? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SelectSaleProductViewHolder {

        val list = LayoutInflater.from(theContext).inflate(
                R.layout.item_sale_product_blob, parent, false)
        return SelectSaleProductViewHolder(list)

    }

    private fun setPictureOnView(imagePath: String, theImage: ImageView) {
        val imageUri = Uri.fromFile(File(imagePath))

        Picasso
                .get()
                .load(imageUri)
                //.resize(0, dpToPxImagePerson())
                .fit()
                .centerCrop()
                .noFade()
                .into(theImage)
    }

    override fun onBindViewHolder(holder: SelectSaleProductViewHolder, position: Int) {

        val entity = getItem(position)
        val imageView = holder.itemView.findViewById<ImageView>(R.id.item_sale_product_blob_image)
        val name = holder.itemView.findViewById<TextView>(R.id.item_sale_product_blob_title)
        val dots = holder.itemView.findViewById<ImageView>(R.id.item_sale_product_blob_dots)

        var imagePathLocal = ""
        var imagePathServer = ""

        productPictureDaoRepo  =
                UmAccountManager.getRepositoryForActiveAccount(theContext).saleProductPictureDao
        productPictureDao = UmAppDatabase.getInstance(theContext).saleProductPictureDao

        holder.imageLoadJob?.cancel()

        imageView.setImageResource(R.drawable.ic_card_giftcard_black_24dp)
        holder.imageLoadJob = GlobalScope.async(Dispatchers.Main) {

            //Load the local image first
            val saleProductPictureLocal = productPictureDao!!.findBySaleProductUidAsync2(
                    entity!!.saleProductUid)
            imagePathLocal = productPictureDaoRepo!!.getAttachmentPath(saleProductPictureLocal!!)!!;

            if (imagePathLocal.isNotEmpty())
                setPictureOnView(imagePathLocal, imageView)
            else
                imageView.setImageResource(R.drawable.ic_card_giftcard_black_24dp)

            //Get the server image
            val saleProductPictureServer =
                    productPictureDaoRepo!!.findBySaleProductUidAsync2(entity.saleProductUid)
            imagePathServer =
                    productPictureDaoRepo!!.getAttachmentPath(saleProductPictureServer!!)!!;

            //If local is not server (suggesting picture/entity update)
            if(saleProductPictureLocal != saleProductPictureServer) {

                if (imagePathServer.isNotEmpty())
                    setPictureOnView(imagePathServer, imageView)
                else
                    imageView.setImageResource(R.drawable.ic_card_giftcard_black_24dp)
            }else{
                //Do nothing
            }
        }



        val currentLocale = impl.getLocale(theContext)
        var saleProductNameLocale: String?=null

        if(currentLocale.equals("fa")){
            saleProductNameLocale = entity!!.saleProductNameDari
        }else if(currentLocale.equals("ps")){
            saleProductNameLocale = entity!!.saleProductNamePashto
        }else{
            saleProductNameLocale = entity!!.saleProductName
        }
        if(saleProductNameLocale == null && entity!!.saleProductName != null) {
            saleProductNameLocale = entity!!.saleProductName

        }
        name.text = saleProductNameLocale


        if (isCatalog) {

            if(entity.saleProductName.equals("Collections")){
                dots.visibility = View.GONE
            }else {
                dots.visibility = View.VISIBLE
            }
            //Options to Edit/Delete every schedule in the list
            dots.setOnClickListener { v: View ->
                if (theActivity != null) {
                    //creating a popup menu
                    val popup = PopupMenu(theActivity!!.applicationContext, v)
                    popup.setOnMenuItemClickListener { item ->
                        val i = item.itemId
                        if (i == R.id.edit) {
                            mPresenter.handleClickProductMulti(entity.saleProductUid, listCategory,
                                    true)
                            true
                        } else if (i == R.id.delete) {
                            mPresenter.handleDelteSaleProduct(entity.saleProductUid, listCategory)
                            true
                        } else {
                            false
                        }
                    }

                    //inflating menu from xml resource
                    popup.inflate(R.menu.menu_edit_delete)
                    popup.menu.findItem(R.id.edit).isVisible = true
                    //If the category is called Collections then hide the delete button
                    if(entity.saleProductName.equals("Collections")){
                        dots.visibility = View.GONE
                    }else {
                        //displaying the popup
                        popup.show()
                    }
                } else if (theFragment != null) {
                    //creating a popup menu
                    val popup = PopupMenu(theFragment!!.context, v)
                    popup.setOnMenuItemClickListener { item ->
                        val i = item.itemId
                        if (i == R.id.edit) {
                            mPresenter.handleClickProductMulti(entity.saleProductUid, listCategory,
                                    true)
                            true
                        } else if (i == R.id.delete) {
                            mPresenter.handleDelteSaleProduct(entity.saleProductUid, listCategory)
                            true
                        } else {
                            false
                        }
                    }

                    //inflating menu from xml resource
                    popup.inflate(R.menu.menu_edit_delete)
                    popup.menu.findItem(R.id.edit).isVisible = true
                    //If the category is called Collections then hide the delete button
                    if(entity.saleProductName.equals("Collections")){
                        dots.visibility = View.GONE
                    }else {
                        //displaying the popup
                        popup.show()
                    }
                }

            }
        } else {
            dots.visibility = View.GONE
        }

        if(mPresenter.isLoggedInPersonAdmin()){
            dots.visibility = View.VISIBLE
        }else{
            dots.visibility = View.GONE
        }

        holder.itemView.setOnClickListener { mPresenter.handleClickProduct(entity.saleProductUid,
                listCategory) }
    }

    inner class SelectSaleProductViewHolder(itemView: View, var imageLoadJob: Job? = null)
        : RecyclerView.ViewHolder(itemView)

    internal constructor(
            diffCallback: DiffUtil.ItemCallback<SaleProduct>,
            thePresenter: SelectSaleProductPresenter,
            activity: Activity,
            isCategory: Boolean?,
            catalog: Boolean?,
            context: Context) : super(diffCallback) {
        mPresenter = thePresenter
        theActivity = activity
        listCategory = isCategory!!
        isCatalog = catalog!!
        theContext = context
    }

    internal constructor(
            diffCallback: DiffUtil.ItemCallback<SaleProduct>,
            thePresenter: SelectSaleProductPresenter,
            fragment: Fragment,
            isCategory: Boolean?,
            catalog: Boolean?,
            context: Context) : super(diffCallback) {
        mPresenter = thePresenter
        theFragment = fragment
        listCategory = isCategory!!
        isCatalog = catalog!!
        theContext = context
    }

    companion object {

        private val IMAGE_WIDTH = 100

        private fun dpToPxImagePerson(): Int {
            return (IMAGE_WIDTH * Resources.getSystem().displayMetrics.density).toInt()
        }
    }


}
